File: final_assignment_backend_quarkus/src/main/java/finalassignmentbackend/service/RefreshTokenService.java
package finalassignmentbackend.service;

import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper;
import com.baomidou.mybatisplus.core.conditions.update.UpdateWrapper;
import finalassignmentbackend.entity.RefreshToken;
import finalassignmentbackend.mapper.RefreshTokenMapper;
import io.quarkus.security.AuthenticationFailedException;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.transaction.Transactional;
import org.eclipse.microprofile.config.inject.ConfigProperty;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.time.LocalDateTime;
import java.util.Base64;
import java.util.HexFormat;
import java.util.List;

/**
 * 刷新令牌服务：签发 / 校验 / 轮换 / 撤销。令牌以 ML-KEM 信封（{@link PqcTokenCrypto}）
 * 后量子加密后落库，明文只在签发时返回给客户端。对齐 Spring service/auth/RefreshTokenService。
 *
 * <p>查找优化：存储时额外写 {@code lookup_digest}（HMAC-SHA-256(raw)），校验/轮换先按 digest 做 O(1)
 * 单行查询；历史无 digest 的令牌回落到全表扫描并回填 digest（一次性迁移）。
 */
@ApplicationScoped
public class RefreshTokenService {

    private static final SecureRandom SECURE_RANDOM = new SecureRandom();
    private static final int REFRESH_TOKEN_BYTES = 32;
    private static final String HMAC_ALGORITHM = "HmacSHA256";
    private static final String HMAC_KEY = "refresh-token-lookup-v1";

    @Inject
    RefreshTokenMapper refreshTokenMapper;

    @Inject
    PqcTokenCrypto pqcTokenCrypto;

    @ConfigProperty(name = "jwt.refresh-token-expiration", defaultValue = "604800")
    long refreshExpirationSeconds;

    @Transactional
    public String createRefreshToken(Long userId) {
        if (userId == null) {
            throw new IllegalArgumentException("userId must not be null");
        }
        if (refreshExpirationSeconds <= 0) {
            throw new IllegalStateException("jwt.refresh-token-expiration must be greater than 0 seconds");
        }

        String raw = generateRawToken();
        LocalDateTime now = LocalDateTime.now();

        RefreshToken entity = new RefreshToken();
        entity.setToken(pqcTokenCrypto.encrypt(raw));
        entity.setLookupDigest(computeDigest(raw));
        entity.setUserId(userId);
        entity.setExpiresAt(now.plusSeconds(refreshExpirationSeconds));
        entity.setRevoked(false);
        entity.setCreatedAt(now);
        refreshTokenMapper.insert(entity);
        return raw;
    }

    @Transactional
    public Long validateRefreshToken(String raw) {
        RefreshToken token = requireActiveToken(raw);
        return token.getUserId();
    }

    @Transactional
    public String rotateRefreshToken(Long userId, String raw) {
        RefreshToken existing = requireActiveToken(raw);
        if (!existing.getUserId().equals(userId)) {
            throw new AuthenticationFailedException("Invalid refresh token");
        }

        // Use optimistic locking via id + revoked=false to handle concurrent rotate
        UpdateWrapper<RefreshToken> update = new UpdateWrapper<>();
        update.eq("id", existing.getId())
                .eq("revoked", false)
                .set("revoked", true);
        int rows = refreshTokenMapper.update(null, update);
        if (rows == 0) {
            throw new AuthenticationFailedException("Refresh token has already been used");
        }

        return createRefreshToken(userId);
    }

    @Transactional
    public void revokeUserTokens(Long userId) {
        if (userId == null) {
            return;
        }
        UpdateWrapper<RefreshToken> update = new UpdateWrapper<>();
        update.eq("user_id", userId)
                .eq("revoked", false)
                .set("revoked", true);
        refreshTokenMapper.update(null, update);
    }

    public long getRefreshTokenExpirationSeconds() {
        return refreshExpirationSeconds;
    }

    private RefreshToken requireActiveToken(String raw) {
        if (raw == null || raw.isBlank()) {
            throw new AuthenticationFailedException("Refresh token is required");
        }

        String digest = computeDigest(raw);

        // O(1) lookup by digest - single row query instead of full table scan
        QueryWrapper<RefreshToken> query = new QueryWrapper<>();
        query.eq("lookup_digest", digest)
                .eq("revoked", false)
                .gt("expires_at", LocalDateTime.now());
        RefreshToken token = refreshTokenMapper.selectOne(query);

        if (token == null) {
            // Fallback: try legacy lookup without digest (for tokens created before migration)
            token = legacyLookup(raw);
            if (token == null) {
                throw new AuthenticationFailedException("Invalid refresh token");
            }
        }

        // Constant-time comparison
        String decrypted;
        try {
            decrypted = pqcTokenCrypto.decrypt(token.getToken());
        } catch (Exception ex) {
            throw new AuthenticationFailedException("Invalid refresh token");
        }

        if (!pqcTokenCrypto.constantTimeEquals(raw, decrypted)) {
            throw new AuthenticationFailedException("Invalid refresh token");
        }

        return token;
    }

    /**
     * Legacy fallback: scan all active tokens (one-time migration support).
     * This should not be called for new tokens that have lookup_digest.
     */
    private RefreshToken legacyLookup(String raw) {
        QueryWrapper<RefreshToken> query = new QueryWrapper<>();
        query.eq("revoked", false)
                .gt("expires_at", LocalDateTime.now())
                .isNull("lookup_digest")
                .last("LIMIT 100");
        List<RefreshToken> candidates = refreshTokenMapper.selectList(query);

        for (RefreshToken candidate : candidates) {
            String decrypted;
            try {
                decrypted = pqcTokenCrypto.decrypt(candidate.getToken());
            } catch (Exception ex) {
                continue;
            }
            if (pqcTokenCrypto.constantTimeEquals(raw, decrypted)) {
                // Backfill the digest for future O(1) lookups
                candidate.setLookupDigest(computeDigest(raw));
                refreshTokenMapper.updateById(candidate);
                return candidate;
            }
        }
        return null;
    }

    private String computeDigest(String raw) {
        try {
            Mac mac = Mac.getInstance(HMAC_ALGORITHM);
            SecretKeySpec keySpec = new SecretKeySpec(
                    HMAC_KEY.getBytes(StandardCharsets.UTF_8), HMAC_ALGORITHM);
            mac.init(keySpec);
            return HexFormat.of().formatHex(mac.doFinal(raw.getBytes(StandardCharsets.UTF_8)));
        } catch (Exception ex) {
            throw new IllegalStateException("Failed to compute refresh token digest", ex);
        }
    }

    private String generateRawToken() {
        byte[] bytes = new byte[REFRESH_TOKEN_BYTES];
        SECURE_RANDOM.nextBytes(bytes);
        return Base64.getUrlEncoder().withoutPadding().encodeToString(bytes);
    }
}

File: final_assignment_backend_go/project/internal/service/refresh_token_service.go
package service

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/subtle"
	"encoding/base64"
	"encoding/hex"
	"errors"
	"time"

	"final_assignment_backend_go/project/internal/auth"
	"final_assignment_backend_go/project/internal/domain"
	"final_assignment_backend_go/project/internal/repo"

	"gorm.io/gorm"
)

// refresh-token-lookup-v1 与 Spring/Quarkus 的 HMAC_KEY 一致，以便跨后端共用 digest 语义。
const refreshHMACKey = "refresh-token-lookup-v1"

var (
	errRefreshTokenRequired = errors.New("refresh token is required")
	errInvalidRefreshToken  = errors.New("invalid refresh token")
	errRefreshAlreadyUsed   = errors.New("refresh token has already been used")
)

// RefreshTokenService 签发 / 校验 / 轮换 / 撤销刷新令牌。令牌以 ML-KEM 信封
// （*auth.PqcTokenCrypto）后量子加密后落库，明文只在签发时返回给客户端。
// 查找优化：存储时额外写 lookup_digest（HMAC-SHA-256(raw)），校验/轮换先按 digest 做 O(1)
// 单行查询；历史无 digest 的令牌回落到全表扫描并回填 digest（一次性迁移）。
// 对齐 Spring/Quarkus 的 RefreshTokenService。
type RefreshTokenService struct {
	repo                     *repo.RefreshTokenRepo
	crypto                   *auth.PqcTokenCrypto
	refreshExpirationSeconds int64
}

// NewRefreshTokenService 构造 RefreshTokenService。refreshExpirationSeconds<=0 时默认 7 天。
func NewRefreshTokenService(r *repo.RefreshTokenRepo, c *auth.PqcTokenCrypto, refreshExpirationSeconds int64) *RefreshTokenService {
	if refreshExpirationSeconds <= 0 {
		refreshExpirationSeconds = 7 * 24 * 3600
	}
	return &RefreshTokenService{repo: r, crypto: c, refreshExpirationSeconds: refreshExpirationSeconds}
}

// CreateRefreshToken 为指定用户签发刷新令牌，返回明文令牌（仅此一次）。
func (s *RefreshTokenService) CreateRefreshToken(userID uint64) (string, error) {
	if userID == 0 {
		return "", errors.New("userId must not be zero")
	}
	return s.createRefreshTokenFor(userID)
}

// createRefreshTokenFor 实际签发逻辑：生成明文、ML-KEM 加密、构造实体并落库。
func (s *RefreshTokenService) createRefreshTokenFor(userID uint64) (string, error) {
	raw := generateRawToken()
	now := time.Now()
	enc, err := s.crypto.Encrypt(raw)
	if err != nil {
		return "", err
	}
	entity := &domain.RefreshToken{
		Token:        enc,
		LookupDigest: s.computeDigest(raw),
		UserID:       userID,
		ExpiresAt:    now.Add(time.Duration(s.refreshExpirationSeconds) * time.Second),
		Revoked:      false,
		CreatedAt:    now,
	}
	if err := s.repo.Insert(entity); err != nil {
		return "", err
	}
	return raw, nil
}

// ValidateRefreshToken 校验原始令牌，返回关联用户 ID。
func (s *RefreshTokenService) ValidateRefreshToken(raw string) (uint64, error) {
	t, err := s.requireActiveToken(raw)
	if err != nil {
		return 0, err
	}
	return t.UserID, nil
}

// RotateRefreshToken 用一次即换：撤销旧令牌并签发新令牌。并发场景靠乐观锁保证单次消费。
// 撤销与签发在同一事务内完成：若签发失败（DB 抖动 / ML-KEM 加密失败 / 唯一索引冲突），
// 撤销随之回滚，避免旧令牌已作废但新令牌未签发导致用户被静默强制登出。对齐 Spring/Quarkus
// 的 @Transactional rotateRefreshToken 语义。
func (s *RefreshTokenService) RotateRefreshToken(userID uint64, raw string) (string, error) {
	existing, err := s.requireActiveToken(raw)
	if err != nil {
		return "", err
	}
	if existing.UserID != userID {
		return "", errInvalidRefreshToken
	}

	var newRaw string
	txErr := s.repo.DB().Transaction(func(tx *gorm.DB) error {
		rows, rerr := s.repo.RevokeByIDTx(tx, existing.ID)
		if rerr != nil {
			return rerr
		}
		if rows == 0 {
			return errRefreshAlreadyUsed
		}
		// 在事务内签发新令牌（加密在事务外亦可，但放事务内便于整体回滚）。
		raw, cerr := s.createRefreshTokenFor(userID)
		if cerr != nil {
			return cerr
		}
		newRaw = raw
		return nil
	})
	if txErr != nil {
		return "", txErr
	}
	return newRaw, nil
}

// RevokeUserTokens 撤销某用户全部未撤销的刷新令牌（登出时调用）。
func (s *RefreshTokenService) RevokeUserTokens(userID uint64) error {
	if userID == 0 {
		return nil
	}
	return s.repo.RevokeByUser(userID)
}

// GetRefreshTokenExpirationSeconds 返回 refresh token 有效期（秒）。
func (s *RefreshTokenService) GetRefreshTokenExpirationSeconds() int64 {
	return s.refreshExpirationSeconds
}

func (s *RefreshTokenService) requireActiveToken(raw string) (*domain.RefreshToken, error) {
	if raw == "" {
		return nil, errRefreshTokenRequired
	}
	digest := s.computeDigest(raw)

	// O(1) lookup by digest
	t, err := s.repo.FindByDigest(digest)
	if err == nil && t != nil && t.ID != 0 {
		if decrypted, derr := s.crypto.Decrypt(t.Token); derr == nil &&
			s.crypto.ConstantTimeEquals(raw, decrypted) {
			return t, nil
		}
	}

	// Fallback: legacy lookup without digest (one-time migration support)
	candidates, lerr := s.repo.FindLegacyCandidates(100)
	if lerr != nil {
		return nil, errInvalidRefreshToken
	}
	for _, candidate := range candidates {
		decrypted, derr := s.crypto.Decrypt(candidate.Token)
		if derr != nil {
			continue
		}
		if subtle.ConstantTimeCompare([]byte(raw), []byte(decrypted)) != 1 {
			continue
		}
		// Backfill the digest for future O(1) lookups
		candidate.LookupDigest = digest
		_ = s.repo.Update(&candidate)
		return &candidate, nil
	}
	return nil, errInvalidRefreshToken
}

func (s *RefreshTokenService) computeDigest(raw string) string {
	mac := hmac.New(newSHA256, []byte(refreshHMACKey))
	mac.Write([]byte(raw))
	return hex.EncodeToString(mac.Sum(nil))
}

func generateRawToken() string {
	b := make([]byte, 32)
	_, _ = rand.Read(b)
	return base64.URLEncoding.WithPadding(base64.NoPadding).EncodeToString(b)
}
